### PR TITLE
565 lab 12 create repair

### PR DESCRIPTION
**Pull request by dm9965**
This Pull Request is for the verification, testing, and approval of both repair sections for Lab 12.

To test the Form Repair, run the ALL application and enter the following url:
_localhost:3000/Lab12/Exercise/FormRepair_

To test the Database Repair, run the ALL application and enter the following url:
_localhost:3000/Lab12/Exercise/DatabaseRepair_

- [ ] No discrepancies across browsers. Eg. chrome vs safari
- [ ] Pages pass W3C Validation for HTML (https://validator.w3.org)
- [ ] CSS (https://jigsaw.w3.org/css-validator/)
- [ ] Google sign-in is operational
- [ ] Accessibility functions
- [ ] Pages can scale without distorting page
- [ ] No dead links
- [ ] Navbar is consistent across the site
- [ ] Pages are screen-reader accessible
- [ ] Contrast meets standards for accessibility
- [ ] Pages are keyboard accessible
- [ ] Passes WAVE Evaluation Tool chrome extension

### CODE REVIEW

**Review by AndreasLc1103 — CHANGES_REQUESTED** [review]
Dev lab 12 needs dev rebase

File: client/src/components/all-components/CodeBlock/Constants.js [hunk 1]
```diff
@@ -1,23 +1,31 @@
...
 
 /* htmlElementRegex uses regex html attributes such as className, onClick, etc */
 export const htmlElementRegex =
-  /\b(className|onClick|disabled|next|fileName|)\b/g;
+  /\b(className|onClick|disabled|next|fileName)\b/g;
+
+export const sqlKeywordRegex =
+  /\b(CREATE|TABLE|COUNT|\*|ALTER|DEFAULT|DELETE|DROP|COLUMN|ROW|FOREIGN KEY|IN|INDEX|INNER JOIN|OUTER JOIN|INSERT INTO|IS NOT NULL|IS NULL| LIMIT|PRIMARY KEY|SELECT|SELECT ALL)\b/g;
+
+export const sqlVariableRegex =
```
Comment (AndreasLc1103) [inline client/src/components/all-components/CodeBlock/Constants.js]: export them at the bottom like this 
`
export {
htmlTagRegex,
etc...
}
`

File: client/src/components/body/Reading/Reading.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ const Reading = (props) => {
   const [readingData, setReadingData] = useState("");
   useScroll();
   useEffect(() => {
+    console.warn(user.userid);
     UserLabService.complete_reading(labID);
```
Comment (AndreasLc1103) [inline client/src/components/body/Reading/Reading.js]: remove

File: client/src/components/body/Reading/Reading.js [hunk 3]
```diff
@@ -17,6 +17,7 @@ const Reading = (props) => {
   const [readingData, setReadingData] = useState("");
   useScroll();
   useEffect(() => {
+    console.warn(user.userid);
```
Comment (heathermoses) [inline client/src/components/body/Reading/Reading.js]: nit: remove console log

**Review by heathermoses — APPROVED** [review]
Tested this by going through both repairs and ensuring error handling worked. LGTM! One small comment to remove console log

File: client/src/components/exercise/lab12/pages/DatabaseRepair.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+} from "../../../../constants/lab12";
+import React, { useEffect } from "react";
+import { EXERCISE_PLAYING } from "../../../../constants";
+import Repair from "../../../body/Repair/Repair";
+import { navigate } from "@reach/router";
+import PropTypes from "prop-types";
+import DatabaseRepairImplementation from "./repairs/DatabaseRepairImplementation";
+import IdentityDatabaseData from "../../../../constants/lab12/DatabaseRepair";
+
+const DatabaseRepair = () => {
```
Comment (AndreasLc1103) [inline client/src/components/exercise/lab12/pages/DatabaseRepair.js:23]: Doc strings

File: client/src/components/exercise/lab12/pages/FormRepair.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  FORM_REPAIR_HEADING,
+  EXERCISE_STATES,
+  REPAIR,
+  EXERCISE_PATH,
+} from "src/constants/lab12/index";
+import FormRepairImplementation from "./repairs/FormRepairImplementation";
+import useMainStateContext from "src/reducers/MainContext";
+import { EXERCISE_PLAYING } from "src/constants/index";
+import IdentityFormData from "../../../../constants/lab12/FormRepair";
+
```
Comment (AndreasLc1103) [inline client/src/components/exercise/lab12/pages/FormRepair.js:15]: Doc strings

File: client/src/components/exercise/lab12/pages/repairs/DatabaseRepairImplementation.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
+import React from "react";
```
Comment (AndreasLc1103) [inline client/src/components/exercise/lab12/pages/repairs/DatabaseRepairImplementation.js:1]: Doc strings describing intent

File: client/src/components/exercise/lab12/pages/repairs/FormRepairImplementation.js [hunk 7]
```diff
@@ -0,0 +1,248 @@
+import React from "react";
```
Comment (AndreasLc1103) [inline client/src/components/exercise/lab12/pages/repairs/FormRepairImplementation.js:1]: same as last comment

File: package.json [hunk 8]
```diff
@@ -20,6 +21,7 @@
     "react-highlight": "^0.15.0",
     "react-live": "^4.0.0",
     "react-syntax-highlighter": "^15.5.0",
+    "react-test-renderer": "^18.2.0",
```
Comment (AndreasLc1103) [inline package.json]: move this to the dev dependencies

File: server/controllers/lab12/ExerciseController.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
+const ExerciseService = require('../../services/lab12/ExerciseService');
```
Comment (AndreasLc1103) [inline server/controllers/lab12/ExerciseController.js:1]: DON'T COPY AND PASTE AND SAY THIS IS DONE

File: server/controllers/lab12/ExerciseController.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ * of the body of the request and pass it off to the Exercise service.
+ * @param {Object} req request object containing user information and payload.
+ */
+async function postExercise(req) {
+  try {
+    const {userID, isAddressComplete, isDateComplete, isNavComplete,
+      isExerciseComplete, hasViewed} = req.body;
+    const responseId = await ExerciseService.postExercise({
+      userId: userID,
+      isAddressComplete: isAddressComplete,
```
Comment (AndreasLc1103) [inline server/controllers/lab12/ExerciseController.js]: CHANGE THESE

File: server/database/models/lab12/Repair.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable new-cap */
+// eslint-disable-next-line max-len
```
Comment (AndreasLc1103) [inline server/database/models/lab12/Repair.js]: remove esflags

**Review by AndreasLc1103 — CHANGES_REQUESTED** [review]
This needs another once over. Caught some moments of copy paste and if you don't fix them that they will become a problem. Also I think you need server test to validate things working so this needs more work.

**Comment by AndreasLc1103** [comment]
Also merge in dev to this branch and the target branch

File: client/src/components/body/Reading/Reading.js (hunk 2 above)
Comment (dm9965) [reply client/src/components/body/Reading/Reading.js]: Done

File: client/src/components/all-components/CodeBlock/Constants.js (hunk 1 above)
Comment (dm9965) [reply client/src/components/all-components/CodeBlock/Constants.js]: done

**Comment by AndreasLc1103** [comment]
Screenshot 2024-04-22 at 1 28 17 PM

File: server/controllers/lab12/ExerciseController.js (hunk 10 above)
Comment (dm9965) [reply server/controllers/lab12/ExerciseController.js]: done

File: server/database/models/lab12/Repair.js (hunk 11 above)
Comment (dm9965) [reply server/database/models/lab12/Repair.js]: done

File: package.json (hunk 8 above)
Comment (dm9965) [reply package.json]: done

File: client/src/components/exercise/lab12/pages/repairs/FormRepairImplementation.js (hunk 7 above)
Comment (dm9965) [reply client/src/components/exercise/lab12/pages/repairs/FormRepairImplementation.js:1]: done

File: client/src/components/exercise/lab12/pages/repairs/DatabaseRepairImplementation.js (hunk 6 above)
Comment (dm9965) [reply client/src/components/exercise/lab12/pages/repairs/DatabaseRepairImplementation.js:1]: done

File: client/src/components/exercise/lab12/pages/FormRepair.js (hunk 5 above)
Comment (dm9965) [reply client/src/components/exercise/lab12/pages/FormRepair.js:15]: done

File: client/src/components/exercise/lab12/pages/DatabaseRepair.js (hunk 4 above)
Comment (dm9965) [reply client/src/components/exercise/lab12/pages/DatabaseRepair.js:23]: done

File: server/controllers/lab12/RepairController.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * submitChange(): is a function that is responsible for
+ * handling when a user submits their change to the address repair fields
+ * @param {Object} req
+ * @param {Object} res
+ */
+async function submitChange(req, res) {
+  const {userID, repair, isComplete, section} = req.body;
+  if (userID === '' || section === '' || repair === '') {
```
Comment (AndreasLc1103) [inline server/controllers/lab12/RepairController.js]: you could also do something like this 
`if (!userID || !section || !repair)`

File: server/controllers/lab12/RepairController.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+/**
+ * getRepair(): is a function that is responsible
+ * for handling retrieving the last recorded
+ * state of a users repair when request
+ * @param {Object} req request object sent to retrieve repair
+ * information.
+ */
+async function getRepair(req) {
```
Comment (AndreasLc1103) [inline server/controllers/lab12/RepairController.js:32]: add in an error case if there is nothing received

**Review by AndreasLc1103 — CHANGES_REQUESTED** [review]
Needs some more work but nothing crazy

**Pull request merged**